### PR TITLE
Fixing race in FSTWriteStream

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This should address https://b.corp.google.com/issues/69701162

There are two races that this PR addresses:

1. The one I found during local testing is that we call `notifyStreamInterruptedWithError` in FSTStream before we clear the stream delegate. It is then possible that FSTRemoteStore re-opens the write stream right away (synchronously). The assert in `startWithDelegate` that verifies that no delegate is set triggers in this case.

2. The other one was root-caused by banging my head on the keyboard a couple times. I believe that it is possible to corrupt the stream state if the idle callback gets schedule while other operations are active. The following sequence should trigger this:

  * Stream goes idle, we schedule the idle callback for execution in 60s
  * 59.9 seconds later, the backend closes the stream for whatever reason. `writeStreamFinished` is invoked and schedules`handleStreamClose` on the dispatch queue.
  * The idle callback triggers and sets the stream state to `Initial`
  * `handleStreamClose` runs and asserts since the stream is no longer active.

I tried to find docs (or sources) for Apple's dispatch queue to verify how events are queued. I was hoping that the guarantees where similar to JS: `setImmediate()` always get priority over `setTimeout()`. This is however not the case - using a small test app I was able to verify that iOS does not prioritize existing `dispatch_async` over `dispatch_after(now)`.

Fixes: https://github.com/firebase/firebase-ios-sdk/issues/490

### CODE REVIEW

File: Firestore/Source/Remote/FSTStream.m [hunk 1]
```diff
@@ -347,6 +347,10 @@ - (void)closeWithFinalState:(FSTStreamState)finalState error:(nullable NSError *
   [self.workerDispatchQueue verifyIsCurrentQueue];
   [self cancelIdleCheck];
 
+  FSTAssert(self.delegate,
```
Comment (wilhuff) [inline Firestore/Source/Remote/FSTStream.m]: Could this move up to live alongside the other assertion, just to group all the sanity checking together?

File: Firestore/Source/Util/FSTDispatchQueue.m [hunk 2]
```diff
@@ -58,7 +58,10 @@ - (void)dispatchAsyncAllowingSameQueue:(void (^)(void))block {
 
 - (void)dispatchAfterDelay:(NSTimeInterval)delay block:(void (^)(void))block {
   dispatch_time_t delayNs = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(delay * NSEC_PER_SEC));
-  dispatch_after(delayNs, self.queue, block);
+  dispatch_after(delayNs, self.queue, ^() {
+    // Make sure that we prioritize tasks that are already queued for immediate execution.
```
Comment (wilhuff) [inline Firestore/Source/Util/FSTDispatchQueue.m]: I don't entirely understand why this double dispatch actually fixes the problem you're describing. There's still a race condition even after this, but it's now just much smaller.

`dispatch_after` still executes on the same serial queue so this block cannot be called concurrently with any other executing block. It doesn't really matter which block executes the close, does it? We still need to deal with the fact after introducing idle timeouts there are now multiple paths to closure and so the `handleStreamClose` [assertion](https://github.com/firebase/firebase-ios-sdk/blob/df76451b3101c47865ccd5721c96a37ca8c8fa70/Firestore/Source/Remote/FSTStream.m#L519) can't be correct.

In short: either the server or the client can close the connection and the event handler that closes it could be either `handleStreamClose` or `handleIdleCloseTimer`. The latter already guards against the possibility that something already closed the stream--we just need to do something similar in `handleStreamClose`.

We need to allow for the possibility that between the point where we get a callback from gRPC indicating close and the point where our own dispatch queue actually handles the server close event `handleIdleCloseTimer` could have run.

File: Firestore/Source/Remote/FSTStream.m (hunk 1 above)
Comment (schmidt-sebastian) [reply Firestore/Source/Remote/FSTStream.m]: Done.

File: Firestore/Source/Util/FSTDispatchQueue.m (hunk 2 above)
Comment (schmidt-sebastian) [reply Firestore/Source/Util/FSTDispatchQueue.m]: Hm, you convinced me - there might be a slim possibility if the idle callback runs while another operation queues a depended block on the queue. So yes, it is better to just let `handleStreamClose` deal with this directly. I originally wanted to keep the state machine cleaner and avoid this extra check.

**Comment by wilhuff** [comment]
Please add a CHANGELOG entry

**Comment by antonioallen** [comment]
Any idea when this fix will be added to the Production Firebase Distribution for iOS? Or which pod do I need to update from this repo to get the fix?

**Comment by wilhuff** [comment]
This will land in the next release, which we're starting to work on assembling now. I can't promise dates because unforeseen circumstances can cause us to miss them, but this should be soon.

If you want to try to pull the pod from source you'll need the pod described in `FirebaseFirestore.podspec`. 

Something like this in your Podfile might get you there (though YMMV since I don't actually test this way):

```
pod 'FirebaseFirebase', :git => 'https://github.com:firebase/firebase-ios-sdk.git', :branch => 'master'
```

Substitute `:commit => '7c7da198eeda7392803e812b0062eb3e9119cade'` if you want this fix but none of the other unreleased progress.

**Comment by wilhuff** [comment]
If you're going to mention specific commits, you probably also want the fix in #522 too: `:commit => '4da1d25ac51ca7236227987b7320fd3cd0f17ab1'`

**Comment by antonioallen** [comment]
Thanks for the reply! So I tried adding the specific commit and I keep getting this error (see image). `pod 'FirebaseFirestore', :git => 'https://github.com:firebase/firebase-ios-sdk.git', :commit => '4da1d25ac51ca7236227987b7320fd3cd0f17ab1'`
error

**Comment by antonioallen** [comment]
Do I need to do something else before I run pod install?

**Comment by morganchen12** [comment]
@antonioallen, you're now in the realm of unsupported experimentation, so our ability to assist you will be limited.

To start debugging your issue I encourage you use `pod install --verbose` to get a more complete error message and then start googling around. The CocoaPods docs are fairly comprehensive and they have a long and extensive list of issues that people have run into in the past on their GitHub, both of which will likely be of use to you.

Good luck!

**Comment by antonioallen** [comment]
Ok, will do thanks! Also, here is a more complete error message (see image). If you have any ideas I would greatly appreciate it! Error: Port number ended with 'f'
error-verbose

**Comment by wilhuff** [comment]
Sorry, that was a typo in the URL. It should have been:

```
https://github.com/firebase/firebase-ios-sdk.git
```

**Comment by antonioallen** [comment]
Aha! That was a sneaky one. However, now the error is a bit more cryptic. Is there a way to do this manually?
error-verbose-2

**Comment by morganchen12** [comment]
There is. CocoaPods' verbose option lists all the commands it runs before running them, so you can just copy paste those commands into your terminal.

**Comment by antonioallen** [comment]
If anyone else encounters this is issue make sure you are on CocoaPods 1.4.0 as specified in the documentation. You can install pre-release versions of Pods by running this command: sudo gem install cocoapods --pre
